File: backend/Backend/src/main/java/com/sgw/backend/entity/MenuItem.java
package com.sgw.backend.entity;

import java.math.BigDecimal;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonIgnore;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
public class MenuItem {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long menuItemId;

    private String menuItemName;

    private String menuItemDescription;

    private String imageURL;

    private BigDecimal sellingPrice;

    private BigDecimal costPrice;

    private boolean enabled;

    private boolean adminBanned;

    /*
     * This is to increase/decrease the price of each line item based on peak hour
     * price multiplier (e.g. 1.2x) OR vocher multiplier (e.g. 0.8x)
     */
    private double voucherMultiplier = 1.0;

    @ManyToOne
    @JsonIgnore
    private MenuSection menuSection;

    @JsonIgnore
    @OneToMany(mappedBy = "menuItem", cascade = CascadeType.ALL)
    List<MenuItemBillable> menuItemBillables;

}

File: student_frontend/lib/model/menu_item.dart
class MenuItem {
  final int menuItemId;
  final String menuItemName;
  final String menuItemDescription;
  String imageURL;
  final double sellingPrice;
  final double costPrice;
  final bool enabled;
  final bool adminBanned;
  final double voucherMultiplier;

  MenuItem({
    required this.menuItemId,
    required this.menuItemName,
    required this.menuItemDescription,
    required this.imageURL,
    required this.sellingPrice,
    required this.costPrice,
    required this.enabled,
    required this.adminBanned,
    required this.voucherMultiplier,
  });

  factory MenuItem.fromJson(Map<String, dynamic> json) {
    return MenuItem(
      menuItemId: json['menuItemId'],
      menuItemName: json['menuItemName'],
      menuItemDescription: json['menuItemDescription'],
      imageURL: json['imageURL'],
      sellingPrice: json['sellingPrice'].toDouble(),
      costPrice: json['costPrice'].toDouble(),
      enabled: json['enabled'],
      adminBanned: json['adminBanned'],
      voucherMultiplier: json['voucherMultiplier'].toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'menuItemId': menuItemId,
      'menuItemName': menuItemName,
      'menuItemDescription': menuItemDescription,
      'imageURL': imageURL,
      'sellingPrice': sellingPrice,
      'costPrice': costPrice,
      'enabled': enabled,
      'adminBanned': adminBanned,
      'voucherMultiplier': voucherMultiplier,
    };
  }

  @override
  String toString() {
    return '''
MenuItem:
  menuItemId: $menuItemId
  menuItemName: $menuItemName
  menuItemDescription: $menuItemDescription
  imageURL: $imageURL
  sellingPrice: $sellingPrice
  costPrice: $costPrice
  enabled: $enabled
  adminBanned: $adminBanned
  voucherMultiplier: $voucherMultiplier
''';
  }
}
